### ISSUE TITLE
Support sending docx files off to Microsoft servers for conversion to PDF

### ISSUE THREAD

**Issue by tbaliance** [baliance/gooxml#185]
## Description

A commercial user mentioned this feature in an anonymous survey, so I'm not sure who told me about it :) Feel free to send me an email or post some information here on how this works.  I'd like to implement it as I think it would be extremely useful.

Thanks!
Todd
todd@baliance.com

**Comment by freb** [comment]
Hey Todd. I just made the repo public so you can take a look:

https://github.com/aerissecure/docx2pdf

As you'll see from the README, the call is extremely simple and requires no authentication.

Another repo I just made public is for converting markdown into docx using blackfriday v2. It could probably use a few cleanups but most everything is supported:

https://github.com/aerissecure/md2docx

I'd like to eventually support fenced code blocks with language identification and a custom Chroma formatter for docx output: https://github.com/alecthomas/chroma

Did you have any thoughts on my suggestion for table of contents generation? As I mentioned, I realize that is a total long shot, but just wondering if its a hard no or something that you think might be feasible at some point.

Thanks!

**Comment by tbaliance** [comment]
@freb I think you're the first person on the internet to discover it, and definitely the first person to document it!

I looked into TOC generation at one point, you can insert the content easily enough but the thing I could never figure out was page number calculations.  You'd have to implement basically the Word layout engine in order to figure out what the page numbers should be.

**Comment by freb** [comment]
That makes total sense about the TOC generation. Even Latex has to make multiple passes for those calculations and its actually doing the rendering.

If the online version of word ever supports TOC generation, there will probably be some way to use it programmatically.

Note that I tried to convert other file formats (xlsx) with the same Microsoft endpoint and it doesn't appear to be supported. OneDrive appears to have an API that can convert between almost anything, but it would never be as simple as the docx->pdf converstion: https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_get_content_format

**Comment by freb** [comment]
Sorry to keep talking about the TOC in this issue. If you think it would be helpful to anyone else, we could start another issue with options and workarounds.

I just figured out it would be possible to host your own conversion server using a real Word installation on Windows using COM. `github.com/go-ole/go-ole` works well for opening and re-saving. Running Windows is far from ideal, but at least there is an option that works and is feasible.

I also found this c# library that generates TOC, but if I understand it correcty they have basically replicated all of the Word rending functionality in their own code for the functions they support, which is nuts. https://help.syncfusion.com/file-formats/docio/overview
